### PR TITLE
Wish/autocomplete car

### PR DESCRIPTION
**Pull request by lenchv**

### CODE REVIEW

File: resources/assets/js/app/services/VehicleService.js [hunk 1]
```diff
@@ -137,10 +138,25 @@ export const VehicleValidate = (data = {
     return result;
 };
 
+export const VehicleData = {
+    fetchBrand(name) {
+        return simpleRequest.get(`/api/v1/brands/${name}/`);
```
Comment (TatsuUkraine) [inline resources/assets/js/app/services/VehicleService.js]: use get params (name or term or similar) to filter brand collection

File: resources/assets/js/app/services/VehicleService.js [hunk 2]
```diff
@@ -137,10 +138,25 @@ export const VehicleValidate = (data = {
...
+    },
+
+    fetchModel(name, brandId) {
+        if (brandId > 0) {
+            return simpleRequest.get(`/api/v1/brands/${brandId}/models/${name}`);
+        } else {
+            return simpleRequest.get(`/api/v1/models/${name}/`);
```
Comment (TatsuUkraine) [inline resources/assets/js/app/services/VehicleService.js]: if brand wasn't selected from autocomplete, don't use autocomplete for models, also add filter Get param to filter collection by name

File: routes/api.php [hunk 3]
```diff
@@ -165,3 +165,13 @@
     'as' => 'reviews.received',
     'uses' => 'ReviewsController@received',
 ]);
+
+Route::get('v1/brands/{name?}/', ['as' => 'brands.byname', 'uses' => 'Api\Car\CarBrandController@brandsByName']);
```
Comment (sergey-ben) [inline routes/api.php]: лучше использовать существующий ендпоинт возвращающий коллекцию брендов - v1/car-brands. Для v1/car-brands стоит добавить обработку get параметра к примеру ?term=ford или ?name=nissan.
Тоже самое касается и ендпоинта моделей. Ну и конечно убрать для этих ендпоинтов миддлвеар аутентифицирующий пользователей

File: app/Http/Controllers/Api/Car/CarBrandController.php [hunk 4]
```diff
@@ -46,4 +49,38 @@ public function getModelByMarkId($id)
...
         }
     }
+
+    /**
+     * @param string $name
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function brandsByName($name = '')
+    {
+        $brands = $this->carBrandService->getByName($name, 100);
```
Comment (sergey-ben) [inline app/Http/Controllers/Api/Car/CarBrandController.php]: лучше использовать константы

File: app/Services/CarBrandService.php [hunk 5]
```diff
@@ -25,4 +26,16 @@ public function getAll()
     {
         return $this->carBrandRepository->all();
     }
+
+    /**
+     * @param string $name
+     * @param int $limit
+     * @return mixed
+     */
+    public function getByName(string $name, int $limit = 10)
```
Comment (sergey-ben) [inline app/Services/CarBrandService.php]: тут тоже лучше константы использовать

**Review by sergey-ben — CHANGES_REQUESTED** [review]

File: resources/assets/js/features/wizard-trip/components/StepThree.js [hunk 6]
```diff
@@ -1,77 +1,162 @@
...
-        this.setState({brand: e.target.value});
+    onBrandChange(e) {
+        const brand = e.target.value,
+            {brands} = this.state;
+
+        this.setState({brand: { value: e.target.value }});
+
+        VehicleData
+            .fetchBrand(brand)
+            .then((response) => this.setState({brands: _.unionBy(brands, response.data.data, 'id')}));
```
Comment (TatsuUkraine) [inline resources/assets/js/features/wizard-trip/components/StepThree.js]: don't store all brands that was fetched, just enough to store last fetched collection

File: resources/assets/js/features/wizard-trip/components/StepThree.js [hunk 7]
```diff
@@ -1,77 +1,162 @@
...
     onModelChange(e) {
-        this.setState({model: e.target.value});
+        const model = e.target.value,
+            { brand, models } = this.state;
+
+        this.setState({model: { value: e.target.value}});
+
+        VehicleData
+            .fetchModel(model, brand.id)
+            .then((response) => this.setState({models: _.unionBy(models, response.data.data, 'id')}));
```
Comment (TatsuUkraine) [inline resources/assets/js/features/wizard-trip/components/StepThree.js]: don't store all models that was fetched, just enough to store last fetched collection

**Review by TatsuUkraine — CHANGES_REQUESTED** [review]

**Review by sergey-ben — DISMISSED** [review]

File: resources/assets/js/app/services/VehicleService.js [hunk 8]
```diff
@@ -137,10 +138,35 @@ export const VehicleValidate = (data = {
...
+    fetchBrand(name) {
+        return simpleRequest.get(`/api/v1/car-brand/`, {
+            params: {
+                search: name,
+                orderBy: 'name'
+            }
+        });
+    },
+
+    fetchModel(name, brandId) {
```
Comment (TatsuUkraine) [inline resources/assets/js/app/services/VehicleService.js]: this method is quite similar to getModelOptions, keep one of them, you can make 1 method that will provide request, and other method will use response to provide different data preparation

File: resources/assets/js/app/services/VehicleService.js [hunk 9]
```diff
@@ -137,10 +138,35 @@ export const VehicleValidate = (data = {
     return result;
 };
 
+export const VehicleData = {
+    fetchBrand(name) {
```
Comment (TatsuUkraine) [inline resources/assets/js/app/services/VehicleService.js]: this method is quite similar to getBrandOptions, keep one of them, you can make 1 method that will provide request, and other method will use response to provide different data preparation

File: resources/assets/js/app/services/VehicleService.js (hunk 8 above)
Comment (TatsuUkraine) [inline resources/assets/js/app/services/VehicleService.js]: if you name method fetch, than it should fetch data with any params, method name should reflect it's logic

**Review by TatsuUkraine — CHANGES_REQUESTED** [review]

**Review by TatsuUkraine — APPROVED** [review]

**Pull request merged**